libgnucash/app-utils/test/gtest-gnc-quotes.cpp

```diff
@@ -0,0 +1,323 @@
+        gnc_commodity_begin_edit(fkcm);
+        gnc_commodity_set_quote_flag(fkcm, TRUE);
+        gnc_commodity_set_quote_source(fkcm, source);
+        gnc_commodity_commit_edit(fkcm);
+        gnc_commodity_table_insert(comm_table, fkcm);
+        gnc_quote_source_set_fq_installed("TestSuite", g_list_prepend(nullptr, (void*)"yahoo_json"));
+    }
+    ~GncQuotesTest() {
+        gnc_clear_current_session();
+    }
+
+    QofSession* m_session;
+    QofBook* m_book;
+};
+
+TEST_F(GncQuotesTest, quote_sources)
+{
+    auto qs_cur{gnc_quote_source_lookup_by_internal("currency")};
+    auto qs_yahoo{gnc_quote_source_lookup_by_internal("yahoo_json")};
+    auto qs_alpha{gnc_quote_source_lookup_by_internal("alphavantage")};
+    EXPECT_TRUE(qs_cur != nullptr);
+    EXPECT_TRUE(qs_yahoo != nullptr);
+    EXPECT_TRUE(qs_alpha != nullptr);
+    EXPECT_TRUE(gnc_quote_source_get_supported(qs_cur));
+    EXPECT_TRUE(gnc_quote_source_get_supported(qs_yahoo));
+    EXPECT_FALSE(gnc_quote_source_get_supported(qs_alpha));
+}
+
+TEST_F(GncQuotesTest, quotable_commodities)
+{
+    auto commodities{gnc_quotes_get_quotable_commodities(gnc_commodity_table_get_table(m_book))};
+    EXPECT_EQ(4u, commodities.size());
+}
+
+#ifdef HAVE_F_Q
+TEST_F(GncQuotesTest, online_wiggle)
+{
+    GncQuotes quotes;
+    quotes.fetch(m_book);
+    auto pricedb{gnc_pricedb_get_db(m_book)};
+    auto failures{quotes.failures()};
+    ASSERT_EQ(1u, failures.size());
+    EXPECT_EQ(GncQuoteError::QUOTE_FAILED, std::get<2>(failures[0]));
+//    EXPECT_EQ(GncQuoteError::QUOTE_FAILED, std::get<2>(failures[1]));
+    EXPECT_EQ(3u, gnc_pricedb_get_num_prices(pricedb));
+}
+#endif
+
+TEST_F(GncQuotesTest, offline_wiggle)
+{
+    StrVec quote_vec{
+        "{\"EUR\":{\"symbol\":\"EUR\",\"currency\":\"USD\",\"success\":\"1\",\"inverted\":0,\"last\":1.0004},\"AAPL\":{\"eps\":6.05,\"success\":1,\"year_range\":\"      129.04 - 182.94\",\"currency\":\"USD\",\"exchange\":\"Sourced from Yahoo Finance (as JSON)\",\"volume\":73539475,\"close\":157.22,\"high\":158.39,\"open\":156.64,\"div_yield\":0.5660857,\"last\":157.96,\"isodate\":\"2022-09-01\",\"method\":\"yahoo_json\",\"name\":\"AAPL (Apple Inc.)\",\"pe\":26.10909,\"low\":154.67,\"type\":\"EQUITY\",\"symbol\":\"AAPL\",\"date\":\"09/01/2022\"},\"HPE\":{\"symbol\":\"HPE\",\"date\":\"09/01/2022\",\"low\":13.13,\"type\":\"EQUITY\",\"method\":\"yahoo_json\",\"name\":\"HPE (Hewlett Packard Enterprise Comp)\",\"isodate\":\"2022-09-01\",\"pe\":4.7921147,\"last\":13.37,\"high\":13.535,\"close\":13.6,\"open\":13.5,\"div_yield\":3.5294116,\"volume\":16370483,\"exchange\":\"Sourced from Yahoo Finance (as JSON)\",\"currency\":\"USD\",\"year_range\":\"        12.4 - 17.76\",\"eps\":2.79,\"success\":1},\"FKCM\":{\"success\":0,\"symbol\":\"FKCM\",\"errormsg\":\"Error retrieving quote for FKCM - no listing for this name found. Please check symbol and the two letter extension (if any)\"}}"
```

Comment:
Can we wrap these long test strings for readability ? I understand this means the string is not EXACTLY the same as what's returned by our F::Q wrapper, but json allows newlines and we're not exactly testing json syntax. We're testing whether our code can properly parse the data encoded in the json...

Reply:
Sure, we can just use C wrapping, e.g.
```
"{\"EUR\":{\"symbol\":\"EUR\",\"currency\":\"USD\",\"success\":\"1\",\"inverted\":0,\"last\":1.0004},\"AAPL\":{\"eps\""
":6.05,\"success\":1,\"year_range\":\"      129.04 - 182.94\",\"currency\":\"USD\",\"exchange\":"
"\"Sourced from Yahoo Finance (as JSON)\",\"volume\":73539475,\"close\":157.22,\"high\":158.39,\"open\":156.64,"
"\"div_yield\":0.5660857,\"last\":157.96,\"isodate\":\"2022-09-01\",\"method\":\"yahoo_json\",\"name\":"
"\"AAPL (Apple Inc.)\",\"pe\":26.10909,\"low\":154.67,\"type\":\"EQUITY\",\"symbol\":\"AAPL\",\"date\":"
"\"09/01/2022\"},\"HPE\":{\"symbol\":\"HPE\",\"date\":\"09/01/2022\",\"low\":13.13,\"type\":\"EQUITY\","
"\"method\":\"yahoo_json\",\"name\":\"HPE (Hewlett Packard Enterprise Comp)\",\"isodate\":\"2022-09-01\","
"\"pe\":4.7921147,\"last\":13.37,\"high\":13.535,\"close\":13.6,\"open\":13.5,\"div_yield\":3.5294116,"
"\"volume\":16370483,\"exchange\":\"Sourced from Yahoo Finance (as JSON)\",\"currency\":"
"\"USD\",\"year_range\":\"        12.4 - 17.76\",\"eps\":2.79,\"success\":1},\"FKCM\":{\"success\":0,\"symbol\":"
"\"FKCM\",\"errormsg\":\"Error retrieving quote for FKCM - no listing for this name found. Please check symbol and"
"the two letter extension (if any)\"}}"
```
Not sure that it improves readability that much, but the compiler removes the newlines so the same string ends up in memory.

Reply:
Well I'd use one line per quote result in that case. That should help with readability. In addition if we ever want to add a quote for an additional test, or change one, it would be easier to spot what's added or changed.